machxo2/arch.cc

```diff
@@ -362,35 +367,127 @@ BelId Arch::getBelByLocation(Loc loc) const
 
 delay_t Arch::estimateDelay(WireId src, WireId dst) const
 {
-    // Taxicab distance multiplied by pipDelay (0.01) and fake wireDelay (0.01).
-    // TODO: This function will not work well for entrance to global routing,
-    // as the entrances are located physically far from the DCCAs.
-    return (abs(dst.location.x - src.location.x) + abs(dst.location.y - src.location.y)) * (0.01 + 0.01);
+    int num_uh = tile_info(dst)->wire_data[dst.index].pips_uphill.size();
+    if (num_uh < 6) {
+        for (auto uh : getPipsUphill(dst)) {
+            if (getPipSrcWire(uh) == src)
+                return getPipDelay(uh).maxDelay();
+        }
+    }
+
+    auto est_location = [&](WireId w) -> std::pair<int, int> {
+        const auto &wire = tile_info(w)->wire_data[w.index];
+        if (w == gsrclk_wire) {
+            auto phys_wire = getPipSrcWire(*(getPipsUphill(w).begin()));
+            return std::make_pair(int(phys_wire.location.x), int(phys_wire.location.y));
+        } else if (wire.bel_pins.size() > 0) {
+            return std::make_pair(wire.bel_pins[0].rel_bel_loc.x,
+                                  wire.bel_pins[0].rel_bel_loc.y);
+        } else if (wire.pips_downhill.size() > 0) {
+            return std::make_pair(wire.pips_downhill[0].rel_loc.x,
+                                  wire.pips_downhill[0].rel_loc.y);
+        } else if (wire.pips_uphill.size() > 0) {
+            return std::make_pair(wire.pips_uphill[0].rel_loc.x,
+                                  wire.pips_uphill[0].rel_loc.y);
+        } else {
+            return std::make_pair(int(w.location.x), int(w.location.y));
+        }
+    };
+
+    auto src_loc = est_location(src);
+    std::pair<int, int> dst_loc;
+    if (wire_loc_overrides.count(dst)) {
+        dst_loc = wire_loc_overrides.at(dst);
+    } else {
+        dst_loc = est_location(dst);
+    }
+
+    int dx = abs(src_loc.first - dst_loc.first), dy = abs(src_loc.second - dst_loc.second);
+
+    return (120 - 22 * device_speed) *
```

Comment:
if you have more trouble routing with this PR, it might be worth just trying increasing the 120 and seeing what happens - as xo2 is slower than ecp5, the delay estimates and predictions may well need to be higher too

Reply:
added higher value, but also made sure that we always use actual speed, so speed 5 is same for XO3 (where only 5 and 6 are available) and for XO2 ( where we have all speeds from 1 to 6)